agent/battle/vision/provider.py

```
@@ -0,0 +1,111 @@
+        self.endpoint = endpoint.rstrip("/")
+        self.api_key = api_key
+        self.model = model
+        self.timeout_s = timeout_s
+        self._urlopen = urlopen or urllib.request.urlopen
+
+    def analyze(self, request: VisionRequest) -> VisionResponse:
+        started = time.monotonic()
+        raw_text = ""
+        try:
+            body = json.dumps(self._payload(request), ensure_ascii=False).encode("utf-8")
+            http_request = urllib.request.Request(self.endpoint, data=body, headers={"Content-Type": "application/json", "Authorization": f"Bearer {self.api_key}"}, method="POST")
```

Ruff (0.16.1)

[error] Audit URL open for permitted schemes. Allowing use of `file:` or custom schemes is often unexpected.

(S310)